blockchain/src/blockchain/zkp_sync.rs

```diff
@@ -0,0 +1,276 @@
+            num_transactions,
+            kind = "push_zkp",
+            "Accepted block",
+        );
+
+        // We shouldn't log errors if there are no listeners
+        _ = this
+            .notifier
+            .send(BlockchainEvent::EpochFinalized(block_hash));
+
+        // We don't have any block logs, so we do not notify the block log stream.
+
+        Ok(PushResult::Extended)
+    }
+
+    /// Pushes a macro block into the blockchain. This is used when we have already synced to the
+    /// most recent election block and now need to push a checkpoint block.
+    /// But this function is general enough to allow pushing any macro block (checkpoint or election)
+    /// at any state of the node (synced, partially synced, not synced).
+    pub fn push_macro(
+        this: RwLockUpgradableReadGuard<Self>,
+        block: Block,
+    ) -> Result<PushResult, PushError> {
+        // Must be a macro block.
+        assert!(block.is_macro());
+
+        // Checks if the body exists.
+        block
+            .body()
+            .ok_or(PushError::InvalidBlock(BlockError::MissingBody))?;
+
+        let read_txn = this.read_transaction();
+
+        // Check if we already know this block.
+        if this
+            .chain_store
+            .get_chain_info(&block.hash(), false, Some(&read_txn))
+            .is_ok()
+        {
+            return Ok(PushResult::Known);
+        }
+
+        if block.block_number() <= this.state.macro_info.head.block_number() {
+            return Ok(PushResult::Ignored);
+        }
+
+        // Perform block intrinsic checks.
+        block.verify(false)?;
+
+        // Verify that the block is a valid macro successor to our current macro head.
+        block.verify_macro_successor(this.state.macro_info.head.unwrap_macro_ref())?;
+
+        // Verify that the block is valid for the current validators.
+        block.verify_validators(&this.current_validators().unwrap())?;
+
+        // At this point we know that the block is correct. We just have to push it.
+
+        // Create the chain info for the new block.
+        let block_hash = block.hash();
+        let block_number = block.block_number();
+        let chain_info = ChainInfo::new(block, true);
+
+        read_txn.close();
+        let mut txn = this.write_transaction();
+
+        let is_election_block = Policy::is_election_block_at(block_number);
+
+        this.chain_store
+            .put_chain_info(&mut txn, &block_hash, &chain_info, true);
+        this.chain_store.set_head(&mut txn, &block_hash);
+
+        // Since it's a macro block, we have to clear the ChainStore.
+        if is_election_block {
+            let max_epochs_stored =
+                cmp::max(this.config.max_epochs_stored, Policy::MIN_EPOCHS_STORED);
+
+            // Calculate the epoch to be pruned. Saturate at zero.
+            let pruned_epoch = Policy::epoch_at(block_number).saturating_sub(max_epochs_stored);
+
+            if this.config.keep_history {
```

Comment:
I think if this logic is only used for a full node, then this could be simplified, since a full node doesn't care about the history

Reply:
We might want to allow full nodes to keep history after syncing as well.